crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -339,23 +334,102 @@ impl SlidingSyncRoom {
+            // If we come from a cold storage, we overwrite the timeline queue with the
+            // timeline updates.
             if self.is_cold.load(Ordering::SeqCst) {
-                // if we come from cold storage, we hard overwrite
-                self.timeline.lock_mut().replace_cloned(timeline);
+                self.timeline_queue.lock_mut().replace_cloned(timeline_updates);
                 self.is_cold.store(false, Ordering::SeqCst);
-            } else if *limited {
-                // the server alerted us that we missed items in between
-                self.timeline.lock_mut().replace_cloned(timeline);
-            } else {
-                let mut ref_timeline = self.timeline.lock_mut();
-                for e in timeline {
-                    ref_timeline.push_cloned(e);
+            }
+            // The server alerted us that we missed items in between.
+            else if *limited {
+                self.timeline_queue.lock_mut().replace_cloned(timeline_updates);
+            }
+            // It's the hot path. We have new updates that must be added to the existing timeline
+            // queue.
+            else {
+                let mut timeline_queue = self.timeline_queue.lock_mut();
+
+                // If the `timeline_queue` contains:
+                //     [D, E, F]
+                // and if the `timeline_updates` contains:
+                //     [A, B, C, D, E, F]
+                // the resulting `timeline_queue` must be:
+                //     [A, B, C, D, E, F]
+                //
+                // To do that, we find the longest suffix between `timeline_queue` and
+                // `timeline_updates`, in this case:
+                //     [D, E, F]
+                // Remove the suffix from `timeline_updates`, we get `[A, B, C]` that is
+                // prepended to `timeline_queue`.
+                //
+                // If the `timeline_queue` contains:
+                //     [A, B, C, D, E, F]
+                // and if the `timeline_updates` contains:
+                //     [D, E, F]
+                // the resulting `timeline_queue` must be:
+                //     [A, B, C, D, E, F]
+                //
+                // To do that, we continue with the longest suffix. In this case, it is:
+                //     [D, E, F]
+                // Remove the suffix from `timeline_updates`, we get `[]`. It's empty, we don't
+                // touch at `timeline_queue`.
+
+                {
+                    let timeline_queue_len = timeline_queue.len();
+                    let timeline_updates_len = timeline_updates.len();
+
+                    let position = match timeline_queue
+                        .iter()
+                        .rev()
+                        .zip(timeline_updates.iter().rev())
+                        .map(|(queue, update)| (queue.event_id(), update.event_id()))
+                        .position(|(queue, update)| queue != update)
```

Comment:
I got really confused here for a second about what is being compared. How about

```suggestion
                        .position(|(queue, update)| queue.event_id() != update.event_id())
```

Reply:
:+1: